src/Microsoft.OData.Edm/Schema/EdmModelBase.cs

```diff
@@ -154,13 +161,31 @@ public IEnumerable<IEdmOperation> FindDeclaredOperations(string qualifiedName)
         /// <returns> A set of operations that share the binding type or empty enumerable if no such operation exists. </returns>
         public virtual IEnumerable<IEdmOperation> FindDeclaredBoundOperations(IEdmType bindingType)
         {
-            foreach (IEnumerable<IEdmOperation> operations in this.functionDictionary.Values.Distinct())
+            IList<IEdmOperation> bindableOperations;
+
+            if (!this.bindableOperationsCache.TryGetValue(bindingType.FullTypeName(), out bindableOperations))
             {
-                foreach (IEdmOperation operation in operations.Where(o => o.HasEquivalentBindingType(bindingType)))
+                HashSet<IList<IEdmOperation>> operationList = new HashSet<IList<IEdmOperation>>();
+                bindableOperations = new List<IEdmOperation>();
+
+                foreach (IList<IEdmOperation> operations in this.functionDictionary.Values)

```

Comment:
you know that this.functionDictionary is an IList, so rather than foreaching over the enumerator, just write a loop that indexes the values by ordinal.